### PR TITLE
ATO-251: Remove parse vtr set logic

### PR DESCRIPTION
**Pull request by hannah-leong**
## What?

- Remove logic from parseVtrSet to only return a list
- Add stringification of vtrList for logging

## Why?

Want to return the list of vtrs, this is a step in doing that

## Related PRs

Branched off from https://github.com/govuk-one-login/authentication-api/pull/3887

### CODE REVIEW

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/VectorOfTrust.java [hunk 1]
```diff
@@ -69,10 +69,31 @@ public static VectorOfTrust parseFromAuthRequestAttribute(List<String> vtr) {
             LOG.warn("Error when parsing vtr attribute", e);
             throw new IllegalArgumentException("Invalid VTR attribute", e);
         }
-        VectorOfTrust vectorOfTrust = parseVtrSet(vtrJsonArray);
-        LOG.info("VTR has been processed at vectorOfTrust: {}", vectorOfTrust);
+        List<VectorOfTrust> vtrList = parseVtrSet(vtrJsonArray);
+        LOG.info("VTR has been processed at vectorOfTrust: [{}]", stringifyVtrList(vtrList));
 
-        return vectorOfTrust;
+        // Todo: return list here
```
Comment (ethanmills) [inline orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/VectorOfTrust.java:75]: Oh?

Comment (hannah-leong) [reply orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/VectorOfTrust.java:75]: Oh.

**Comment by hannah-leong** [comment]
Further steps to have `parseFromAuthRequestAttribute` return a `List<VectorOfTrust` in future PR

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthenticationCallbackHandler.java [hunk 2]
```diff
@@ -486,11 +484,15 @@ private Map<String, String> buildDimensions(
...
+        var orderedVtrList = VectorOfTrust.orderVtrList(clientSession.getVtrList());
+        var mfaRequired =
+                !orderedVtrList
+                        .get(0)
+                        .getCredentialTrustLevel()
+                        .equals(CredentialTrustLevel.LOW_LEVEL);
         var levelOfConfidence = LevelOfConfidence.NONE.getValue();
-        if (vtr.containsLevelOfConfidence()) {
-            levelOfConfidence = vtr.getLevelOfConfidence().getValue();
+        if (orderedVtrList.get(0).containsLevelOfConfidence()) {
```
Comment (ethanmills) [inline oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthenticationCallbackHandler.java:494]: I think we said we were going to write a comment here about the assumption we've made here implicitly, and why it's justified?

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthenticationCallbackHandler.java [hunk 3]
```diff
@@ -486,11 +484,15 @@ private Map<String, String> buildDimensions(
...
+            levelOfConfidence = orderedVtrList.get(0).getLevelOfConfidence().getValue();
```
Comment (ethanmills) [inline oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthenticationCallbackHandler.java:495]: This doesn't look right to me. We spoke about a sorted, comma separated list of the requested values

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/VectorOfTrust.java [hunk 4]
```diff
@@ -69,10 +69,30 @@ public static VectorOfTrust parseFromAuthRequestAttribute(List<String> vtr) {
             LOG.warn("Error when parsing vtr attribute", e);
             throw new IllegalArgumentException("Invalid VTR attribute", e);
         }
-        VectorOfTrust vectorOfTrust = parseVtrSet(vtrJsonArray);
-        LOG.info("VTR has been processed at vectorOfTrust: {}", vectorOfTrust);
+        List<VectorOfTrust> vtrList = parseVtrSet(vtrJsonArray);
+        LOG.info("VTR has been processed at vectorOfTrust: [{}]", stringifyVtrList(vtrList));
 
-        return vectorOfTrust;
+        return vtrList.get(0);
```
Comment (ethanmills) [inline orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/VectorOfTrust.java:75]: We've lost all the sorting and ordering that was in the `parse` method and now just return the first at random?

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/VectorOfTrust.java [hunk 5]
```diff
@@ -69,10 +69,30 @@ public static VectorOfTrust parseFromAuthRequestAttribute(List<String> vtr) {
...
+
+    public static CredentialTrustLevel getLowestCredentialTrustLevel(List<VectorOfTrust> vtrList) {
+        List<VectorOfTrust> orderedVtrList = orderVtrList(vtrList);
+        if (orderedVtrList.isEmpty()) {
+            throw new IllegalArgumentException("Invalid VTR attribute");
+        }
+        return orderedVtrList.get(0).getCredentialTrustLevel();
+    }
+
+    public static List<VectorOfTrust> orderVtrList(List<VectorOfTrust> vtrList) {
```
Comment (ethanmills) [inline orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/VectorOfTrust.java:86]: Are there any tests for this?

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthenticationCallbackHandler.java (hunk 2 above)
Comment (hannah-leong) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthenticationCallbackHandler.java:494]: That comment is in the identity helper. Here we want the `RequestedLevelOfConfidence` which is the lowest requested value (hence ordering and taking the first)

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthenticationCallbackHandler.java (hunk 3 above)
Comment (hannah-leong) [reply oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthenticationCallbackHandler.java:495]: Ahh okay, I put that in the method logging but shall put it here too

File: integration-tests/src/test/java/uk/gov/di/authentication/api/AuthorisationIntegrationTest.java [hunk 6]
```diff
@@ -719,8 +718,7 @@ void shouldCallAuthorizeAsDocAppClient(String uiLocales) throws JOSEException, P
                 authRequest.getCustomParameter("vtr"),
                 equalTo(List.of("[\"P2.Cl.Cm\",\"Cl.Cm\"]")));
```
Comment (hannah-leong) [inline integration-tests/src/test/java/uk/gov/di/authentication/api/AuthorisationIntegrationTest.java:719]: This test includes an invalid vtr list (mix of identity and non-identity) - this should be caught with the updated logic in ato-248

File: integration-tests/src/test/java/uk/gov/di/authentication/api/AuthorisationIntegrationTest.java [hunk 7]
```diff
@@ -109,7 +109,7 @@ class AuthorisationIntegrationTest extends ApiGatewayHandlerIntegrationTest {
     private static final String CLAIMS =
             "{\"userinfo\":{\"https://vocab.account.gov.uk/v1/coreIdentityJWT\":{\"essential\":true},\"https://vocab.account.gov.uk/v1/address\":null}}";
 
-    private static IntegrationTestConfigurationService configuration =
+    private static final IntegrationTestConfigurationService configuration =
```
Comment (ethanmills) [inline integration-tests/src/test/java/uk/gov/di/authentication/api/AuthorisationIntegrationTest.java:112]: You're going to get a conflict with https://github.com/govuk-one-login/authentication-api/pull/3911 on this. This is a good reason not to make unrelated changes in other files 🙃

File: oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthenticationCallbackHandler.java [hunk 8]
```diff
@@ -489,11 +490,13 @@ private Map<String, String> buildDimensions(
...
         var levelOfConfidence = LevelOfConfidence.NONE.getValue();
+        // Assumption: Requested vectors of trust will either all be for identity or none, and so we
+        // can check just the first
         if (orderedVtrList.get(0).containsLevelOfConfidence()) {
-            levelOfConfidence = orderedVtrList.get(0).getLevelOfConfidence().getValue();
+            levelOfConfidence = VectorOfTrust.stringifyLevelsOfConfidence(orderedVtrList);
         }
         dimensions.put("MfaRequired", mfaRequired ? "Yes" : "No");
-        dimensions.put("RequestedLevelOfConfidence", levelOfConfidence);
+        dimensions.put("RequestedLevelOfConfidence", format("[%s]", levelOfConfidence));
```
Comment (ethanmills) [inline oidc-api/src/main/java/uk/gov/di/authentication/oidc/lambda/AuthenticationCallbackHandler.java:499]: Normally I'd support wrapping this in square brackets, but as is we've got loads of legacy data with a single string this will be incompatible with. All the `P2` will now be `[P2]` and the graphs will break, right?

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/VectorOfTrust.java [hunk 9]
```diff
@@ -177,6 +192,14 @@ public static List<String> getRequestedLevelsOfConfidence(List<VectorOfTrust> vt
                 .toList();
     }
 
+    private static String stringifyVtrList(List<VectorOfTrust> vtrList) {
+        var vtrBuilder = new StringBuilder();
+        for (VectorOfTrust vot : vtrList) {
+            vtrBuilder.append(vot.toString()).append(",");
+        }
+        return vtrBuilder.substring(0, vtrBuilder.length() - 1);
```
Comment (ethanmills) [inline orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/VectorOfTrust.java:200]: Are these the same?
```suggestion
        return vtrList.stream()
                .map(VectorOfTrust::toString)
                .collect(Collectors.joining(","));
```

File: orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/VectorOfTrust.java [hunk 10]
```diff
@@ -187,6 +210,14 @@ public String toString() {
                 + '}';
     }
 
+    public static String stringifyLevelsOfConfidence(List<VectorOfTrust> votList) {
+        var votBuilder = new StringBuilder();
+        for (VectorOfTrust vot : votList) {
+            votBuilder.append(vot.getLevelOfConfidence().getValue()).append(",");
+        }
+        return votBuilder.substring(0, votBuilder.length() - 1);
```
Comment (ethanmills) [inline orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/entity/VectorOfTrust.java:218]: Likewise
```suggestion
        return votList.stream()
                .map(VectorOfTrust::getLevelOfConfidence)
                .map(LevelOfConfidence::getValue)
                .collect(Collectors.joining(","));
```

File: integration-tests/src/test/java/uk/gov/di/authentication/api/AuthorisationIntegrationTest.java (hunk 7 above)
Comment (hannah-leong) [reply integration-tests/src/test/java/uk/gov/di/authentication/api/AuthorisationIntegrationTest.java:112]: This was meant to be in a different PR, will move

**Review by ethanmills — APPROVED** [review]
